### PR TITLE
Challenge 2016 Solution in Golang - Thamizh

### PR DESCRIPTION
**Pull request by thamizhv**

### CODE REVIEW

**Comment by maknahar** [comment]
Hi Thamizh,
Thanks for the submission.

Could you please add documentation for the solution? Things like how server need to start, what are the dependencies, API request and response format etc.

Thanks.

**Comment by thamizhv** [comment]
Hi Mayank,

The solution.md file contains the required details. There are two postman collections attached with sample requests (One for admin role, one for distributor role). I have updated the soluiton.md file with action flow details. Please have a look and let me know for any queries.

The server can be started by 
MacOS/Linux
```
go get ./...
go run *.go

```

Windows
```
go get ./...
go run ./

```
When go run command is used server will be started at 8080 port. 
To stop the server press `Cmd + C` or `Ctrl + C`

**Comment by maknahar** [comment]
I tried to start the server but it seems there is some compile time error. Looks like dependency issue to me though.
[
screenshot 2019-02-20 at 11 34 22 pm
](url)

Could you please vendorize your service?

File: authController.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+	if req.Method != "POST" {
+		respondError(w, http.StatusMethodNotAllowed, methodNotAllowed)
+		return
+	}
+
+	creds := credential{}
+
+	err := json.NewDecoder(req.Body).Decode(&creds)
+	defer req.Body.Close()
```
Comment (maknahar) [inline authController.go]: Any reason defer statement is here instead of at the top of the function?

File: locations.go [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+	outCountry.Name = c.Name
+	outCountry.Code = c.Code
+
+	_, p, _ := getProvinceFromCountry(c.Provinces, provinceName)
+	if p == nil {
+		return country{}, errors.New(locationNotFound)
+	}
+
+	if cityName == "" {
+		tempProvince := copyProvince(*p)
```
Comment (maknahar) [inline locations.go]: is there a way to avoid manually creating a copy?

**Comment by thamizhv** [comment]
Hi Mayank,

Yes it is dependency issue. Only two outer packages are used. 

Please run below command as mentioned above.
```go get ./...```
This will install all dependencies.

Or run below commands

```
go get github.com/satori/go.uuid
go get golang.org/x/crypto/bcrypt
```

Please let me know if issue still exists.

**Comment by maknahar** [comment]
Would it be possible to vendorise your code instead of installing the dependencies in GOPATH?

**Comment by thamizhv** [comment]
I have vendorized the dependencies.

File: locations.go (hunk 2 above)
Comment (thamizhv) [reply locations.go]: I tried copy() function but I didn't get deep copied result. So, I manually copied. Can try any deep copy external packages.

File: authController.go (hunk 1 above)
Comment (thamizhv) [reply authController.go]: Yes. It should be above line 20. Here it can be at top of function as req.Body.Close() wont evaluate to nil.

**Pull request closed without merging**